### PR TITLE
add a attack  'alignment' attribute for have specific attack alignment different of unit alignment

### PR DESCRIPTION
**Pull request by newfrenchy83**
it is almost same what https://github.com/wesnoth/wesnoth/pull/6968 who is rebased with this commit but i add only attribute in [attack] here

### CODE REVIEW

**Comment by Pentarctagon** [comment]
Does this supercede #6968 then, or are both needed?

**Comment by newfrenchy83** [comment]
> Does this supercede #6968 then, or are both needed?

Like i said, the commit of this pr is New base to https://github.com/wesnoth/wesnoth/pull/6968, it can be merged apart if idea of weapon special is reject but i prefer both

**Comment by Pentarctagon** [comment]
@knyghtmare given your comment at https://github.com/wesnoth/wesnoth/pull/6968#issuecomment-2101278325, do you have a preference one way or the other?

**Comment by newfrenchy83** [comment]
i modified code for what if 'alignment' write with attack_alignment or unit_type alignment attribute don't matche with regular alignment then "neutral" value was returned for maintain coherent with actual logic

**Comment by knyghtmare** [comment]
> @knyghtmare given your comment at [#6968 (comment)](https://github.com/wesnoth/wesnoth/pull/6968#issuecomment-2101278325), do you have a preference one way or the other?

I would suggest the simplest approach, adding in the new feature without distorting the existing code of UMC/core units.
The example below works:
```ini
[attack]
    attack_alignment=lawful/chaotic/neutral/liminal
[/attack]
```

May I ask if existing units in core/umc will stay the same? or will need `attack_alignment=` in their attacks as obligatory if this feature is merged?

**Comment by newfrenchy83** [comment]
> > @knyghtmare given your comment at [#6968 (comment)](https://github.com/wesnoth/wesnoth/pull/6968#issuecomment-2101278325), do you have a preference one way or the other?
> 
> I would suggest the simplest approach, adding in the new feature without distorting the existing code of UMC/core units. The example below works:
> 
> ```ini
> [attack]
>     attack_alignment=lawful/chaotic/neutral/liminal
> [/attack]
> ```
> 
> May I ask if existing units in core/umc will stay the same? or will need `attack_alignment=` in their attacks as obligatory if this feature is merged?

existing units can stay the same because attack_alignment replace unit alignment only when used, else unit alignment will be used like before

**Comment by knyghtmare** [comment]
What is the difference between this PR #8851 and #6968 ? They seem to come off as kind of the same so want to know what's really happening.

**Comment by newfrenchy83** [comment]
> What is the difference between this PR #8851 and #6968 ? They seem to come off as kind of the same so want to know what's really happening.

in https://github.com/wesnoth/wesnoth/pull/6968 I add a special weapon [attack_alignment] in addition to the attack_alignment attribute which returns in addition to the alignment specified when the special weapon is active a percentage of the tod_bonus (100% by default), this allows you to create types customized similar to lawful or chaotic, but with different values (a half lawful of 13% of tod instead of 25%, or a chaotic of 50% given to an attack of a chaotic unit to make it super-chaotic.
The other advantage is being able to condition the application of the alignment to precise conditions (a bit like a night aim which could break down or which is only usable offensively while its owner is neutral at best).

This PR can be committed immediately and https://github.com/wesnoth/wesnoth/pull/6968 saved for later.

File: data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg
Comment (Pentarctagon) [inline data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:24]: Please add documentation for what this test is doing. Also, why is the Illuminates ability needed?

Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:24]: for what alignment change was tested, Tod of test must be changed for 'lawful' alignment change for what Todbonus was diffrent of ero

Comment (Pentarctagon) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:24]: Please add a test that uses [end_turn] to make it be a lawful ToD as well then.

Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:24]: i don't think what standart Tod was implemente in test files, it's why i use illuminates ability instead

Comment (Pentarctagon) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:24]: For `COMMON_KEEP_A_B_UNIT_TEST` it looks like there's already an optional TIME argument actually.

Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:23]: @Pentarctagon documentation added and illuminate removed

Comment (Pentarctagon) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg]: What I meant was that there should be both tests:
* one test that works due to Illuminates
* a second test that works because of ToD

Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg:62]: done

**Comment by Pentarctagon** [comment]
Added a commit to handle get_enum() potentially returning nullopt. Beyond that I think this should have a test for filtering on the weapon alignment and a test for using in WFL, but otherwise I think it looks good.

**Comment by soliton-** [comment]
I don't think it makes sense to use the `attack_` prefix when no other attribute (besides `attack_weight` where it has a different meaning) uses it. If this is deemed useful (without any motivation I can see) then it should just be called `alignment`.

File: src/units/attack_type.hpp
Comment (soliton-) [inline src/units/attack_type.hpp:360]: This looks like a copy of `unit_alignments`. If this is really needed that should be clearly documented and explained why `values` of the enum cannot be used.

File: src/units/abilities.cpp
Comment (soliton-) [inline src/units/abilities.cpp:1309]: Untested:
```suggestion
		// pick attack alignment or fall back to unit alignment
		return unit_alignments::get_enum(attack_alignment()).value_or(self ? self_->alignment() : unit_alignments::type::neutral);
	}
```

Comment (soliton-) [reply src/units/abilities.cpp:1309]: IMO the function should be called `alignment` just like for the unit type.

**Comment by newfrenchy83** [comment]
> I don't think it makes sense to use the `attack_` prefix when no other attribute (besides `attack_weight` where it has a different meaning) uses it. If this is deemed useful (without any motivation I can see) then it should just be called `alignment`.

i don't want whar wml develloper confuse alignment with attack_alignment when using [effect]

File: src/units/attack_type.hpp
Comment (newfrenchy83) [reply src/units/attack_type.hpp]: removed

**Comment by soliton-** [comment]
> i don't want whar wml develloper confuse alignment with attack_alignment when using [effect]

You mean the confusion would be between
```
[effect]
     apply_to=alignment
     set=chaotic
[/effect]
```
and
```
[effect]
     apply_to=attack
     alignment=chaotic
[/effect]
```
?

**Comment by newfrenchy83** [comment]
> > i don't want whar wml develloper confuse alignment with attack_alignment when using [effect]
> 
> You mean the confusion would be between
> 
> ```
> [effect]
>      apply_to=alignment
>      set=chaotic
> [/effect]
> ```
> 
> and
> 
> ```
> [effect]
>      apply_to=attack
>      set_alignment=chaotic
> [/effect]
> ```
> 
> ?
Yes

**Comment by soliton-** [comment]
Doesn't seem reasonable to me.

**Comment by newfrenchy83** [comment]
@soliton- The tests finally succeed, do you really want to change 'attack_alignment' to 'alignment'? I think keeping two different terms would be more readable for developers, 'alignment' in a filter immediately evoking the alignment of the unit and 'attack_alignment' the alignment encoded specifically in an attack especially if the difference between filter and 'filter_attack' is not immediately detected by those who try to decrypt the wml code.

**Comment by newfrenchy83** [comment]
@Pentarctagon simplified code by suggestion of @soliton-

**Comment by soliton-** [comment]
If alignment is inextricably linked to unit alignment how does this whole PR even make sense (without any motivation on why attack alignment needs to be introduced at that)?

**Comment by Pentarctagon** [comment]
@newfrenchy83 in what cases would in be unclear? Using `alignment` in `[filter_attack]` for example is clear what it means.

**Comment by newfrenchy83** [comment]
> @newfrenchy83 in what cases would in be unclear? Using `alignment` in `[filter_attack]` for example is clear what it means.

OK i will change that to alignment then

**Comment by newfrenchy83** [comment]
@Pentarctagon Using in filters and formulas 'attack_alignment' only detects what is encoded in the attack, if for example unit alignment is 'chaotic' attack_alignment is 'defg' and [filter_attack]attack_alignment=chaotic, the filter will not match because it detects the encoded character string.

If I replace attack_alignment with 'alignment' the developer could believe that in the absence of [attack]alignment= the filter will detect the [unit_type]alignment= which is not the case, I must first resolve this difference before changing the name of the attribute or leave the current name.

**Comment by Pentarctagon** [comment]
I'm not understanding. Can you provide an example where it's unclear?

**Comment by newfrenchy83** [comment]
> I'm not understanding. Can you provide an example where it's unclear?

replaced by 'alignment'

**Comment by Pentarctagon** [comment]
@soliton- any other issues?

File: changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md
Comment (soliton-) [inline changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md:2]: Can't really tell what this says.

File: src/reports.cpp
Comment (soliton-) [inline src/reports.cpp:968]:
```suggestion
			if(tod_bonus != 0) {
				color = (tod_bonus > 0) ? font::good_dmg_color : font::bad_dmg_color;
			}
```

File: src/units/attack_type.hpp
Comment (soliton-) [inline src/units/attack_type.hpp:94]: Why are 3 functions necessary to get the alignment when unit gets along with 1? IMO there should be a single function `alignment` that returns `unit_alignments::type` just like for units.

File: src/units/attack_type.hpp
Comment (soliton-) [inline src/units/attack_type.hpp:367]: IMO this should also be just `alignment_`.

File: src/reports.cpp
Comment (soliton-) [inline src/reports.cpp:789]: Here it would be useful to use weapon_- or attack_alignment as the variable name.

**Review by soliton- — CHANGES_REQUESTED** [review]

**Comment by soliton-** [comment]
> @soliton- any other issues?

Once the review comments are resolved just need to go over all the non-code text (commit message, comments, etc) and fix the language.

File: src/units/attack_type.hpp
Comment (newfrenchy83) [reply src/units/attack_type.hpp:367]: if alignment_ variable exist in unit class i prefer let the actual name.

File: src/units/attack_type.hpp
Comment (newfrenchy83) [reply src/units/attack_type.hpp:94]: This is because the alignment in [attack] must replace the unit alignment when it is correctly entered, otherwise it is the latter that is exercised, and this must be done in a function where the self_ variable is entered, this is why alignment_ in attack_tpe is a string variable. For the filtering check, it is the alignment actually used (if invalid alignment returns unit alignment) that is converted back to a string.

Finally, in the lua, self_ cannot be entered and it is therefore the alignment_ variable that is directly returned. So three functions are necessary

File: changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md
Comment (newfrenchy83) [reply changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md]: changed changelog entry

File: src/reports.cpp
Comment (newfrenchy83) [reply src/reports.cpp]: done

File: src/units/attack_type.hpp
Comment (newfrenchy83) [reply src/units/attack_type.hpp:369]: i prefer be careful and use a diffrent name what for type and unit class

File: src/reports.cpp
Comment (newfrenchy83) [reply src/reports.cpp]: done

File: src/units/attack_type.hpp
Comment (soliton-) [reply src/units/attack_type.hpp:369]: Why is that helpful? An attack also has `id_`, `type_`, `description_`, etc all without an attack_ prefix even though unit has those as well. To me it seems more confusing to special case alignment.

File: data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg]: suggestion aplied

File: data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/UnitsWML/Attacks/attack_alignment.cfg]: suggestion applied

File: changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md
Comment (newfrenchy83) [reply changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md]: partially applied suggestion and removed list of valid alignment

File: changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md
Comment (newfrenchy83) [reply changelog_entries/[unit_alignment]_[attack_alignment]_changelog.md]: removed lua code line

File: src/units/attack_type.hpp
Comment (newfrenchy83) [reply src/units/attack_type.hpp]: changed to alignment_

File: src/units/attack_type.hpp
Comment (soliton-) [reply src/units/attack_type.hpp:96]: I understand that you use all those functions but that does not seem reasonable and quite confusing to figure out from the names which function does what.

Internally you can use `alignment_` if you need the original string and otherwise you call `alignment()` which gives you the actual alignment. If you need the string then get the string from there.

If you really must have 3 different functions that do nearly the same thing they need to have _very_ good names to be able to tell what is what. But since unit gets along fine without that I don't see the point.

Comment (newfrenchy83) [reply src/units/attack_type.hpp]: @soliton- names changed, have you suggestion if these name are not convenient?

Comment (soliton-) [reply src/units/attack_type.hpp]: Since you use `effective_alignment_str` you should also use `alignment_str`. I suppose that's clear enough then.

Comment (soliton-) [reply src/units/attack_type.hpp]: Actually while the names are fine it's still not clear why `alignment_str()` is a useful public API. It seems to be the opposite it invites to hand out possibly invalid alignment strings.

Comment (newfrenchy83) [reply src/units/attack_type.hpp]: fixed with https://github.com/wesnoth/wesnoth/pull/8851/commits/b0fbac1393394babd181efc61f1c7748f96b2d61 commit

Comment (newfrenchy83) [reply src/units/attack_type.hpp]: @soliton- lokk at https://github.com/wesnoth/wesnoth/pull/8851/commits/b0fbac1393394babd181efc61f1c7748f96b2d61

**Comment by newfrenchy83** [comment]
@soliton- rebased this Pr

File: src/units/abilities.cpp
Comment (soliton-) [inline src/units/abilities.cpp:1300]: Can you explain in detail how to see the problem why this should be needed? Perhaps open an issue about it so other people can take a look.

For this PR I suggest to remove this function.

Comment (newfrenchy83) [reply src/units/abilities.cpp:1300]: i don't want attack alignment return neutral when attribute empty in lua code where i can't put unit data

Comment (newfrenchy83) [reply src/units/abilities.cpp:1300]: In the impl_unit_attack_get() function in the lua unit attacks.cpp file the unit information is not present despite all my efforts to solve this problem, which means that if I return a simple reverse transcription of attack alignment in string but alignment_ is empty or badly transcribed then it will be interpreted as a neutral alignment to avoid this while waiting to find a more appropriate solution, in case of absence of self_ and therefore of unit alignment, the string should simply alignment_str_ without selection, so if it is empty, it is empty and not neutral (empty means no alignment and not neutral alignment), when self_ present could return unit alignment if weapon alignment no valid.

Comment (CelticMinstrel) [reply src/units/abilities.cpp:1300]: We can't be doing weird workarounds for the fact that the attack does not have `self_` when you access it from Lua. This is not a bug, exactly. It is a missing feature. What we need is a way for Lua to set the `self_` on an attack; but if the attack for whatever reason doesn't have it set, then it should return the proper data that you'd expect in such a case (probably the base data).

Comment (newfrenchy83) [reply src/units/abilities.cpp:1300]: i all try for resolve problem of uint data in this table without succes, someone can help me?

Comment (newfrenchy83) [reply src/units/abilities.cpp:1300]: for me the proper data in absence of self_ in lua this is the base data (alignment_str_) without modification and i must let the function like it is or use a base_alignment() function for return directly alignment_str_

Comment (newfrenchy83) [reply src/units/abilities.cpp]: > What I recall from what was discussed on discord is that this isn't possible to do.

what it isn't possible to do?

Comment (Pentarctagon) [reply src/units/abilities.cpp]: You can check the IRC logs from the 19th, but:
> I don't think impl_unit_attack_get has a way to access the unit who owns the attack. There might not even be one. It could be an attack belonging to unit type, or a standalone attack.

Comment (soliton-) [reply src/units/abilities.cpp]: Btw, as far as I'm concerned I'm fine with merging this (with a sane interface not in the current state) even if usage from lua does not work right. The fix needs to be done in the lua integration and I'm fairly certain the fix is not just needed for this. (IMO there is a good chance that #9232 and #9206 for example are related.)

Comment (newfrenchy83) [reply src/units/abilities.cpp]: whaat issue has the interface? For lua i don't know how fix this.

Comment (soliton-) [reply src/units/abilities.cpp]: The interface should not have functions that do special logic just for lua. As I'm sure if said multiple times by now IMO there should be a single function that returns the alignment and that's it.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: i changed alignment_str() code for handle lua case, i can't doing best.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: I know, it one alternative is to remove lua code.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: @soliton- if i removed attribute of lua code,in wait of fix o lua integration, will you merge this PR?

Comment (soliton-) [reply src/units/abilities.cpp]: If you mean not allowing lua to access the alignment attribute for now then I guess that should be fine as well.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: removed from lua.

**Comment by CelticMinstrel** [comment]
From what I can tell, there is no real reason for `alignment_str()` to exist.

**Review by soliton- — APPROVED** [review]
The commit message should be:
Add an 'alignment' attribute to attacks

This attribute, when set to a valid value of lawful|neutral|chaotic|liminal, will assign to the weapon a different alignment than the unit alignment used by default. This alignment is then used when attacking with this weapon.

The attribute is not accessible from lua so far since the fallback to unit alignment does not work.

**Comment by soliton-** [comment]
> From what I can tell, there is no real reason for `alignment_str()` to exist.

Yeah, but I don't mind it as a convenience.

**Comment by CelticMinstrel** [comment]
> fall back

fallback

**Comment by newfrenchy83** [comment]
> From what I can tell, there is no real reason for `alignment_str()` to exist.

lighten writing in src/formula/callable_objects.cpp and src/units/attack_type.cpp

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp]: @soliton- @stevecotton resolved issue with https://github.com/wesnoth/wesnoth/pull/8851/commits/f5fd70d16800238bb8d5374fccdef46f71305da4

**Comment by ProditorMagnus** [comment]
https://wiki.wesnoth.org/index.php?title=UnitTypeWML&action=historysubmit&type=revision&diff=73799&oldid=73576

File: src/units/abilities.cpp
Comment (soliton-) [reply src/units/abilities.cpp]: Ok, missed this before merging. Duplicating the existing alignments there does not look good. Oh well...

Comment (newfrenchy83) [reply src/units/abilities.cpp]: This is the only solution that works in Lua and I can't put self_ in attack_type() in all files where it is called because it is not always called with an attack_type variable, this hybrid solution is the only one that works and I need that when attack_type() does not have self_ and alignment is empty or invalid that alignment_str_ is empty and not replaced by a neutral alignment so that alignment() can return the unit alignment instead of attack_type()

File: src/units/attack_type.cpp
Comment (soliton-) [inline src/units/attack_type.cpp:106]: What is this check for? Why set weapon alignment based on `self_`?

File: src/units/abilities.cpp
Comment (soliton-) [reply src/units/abilities.cpp]: My point was that you duplicated the alignment strings instead of using `unit_alignments::get_enum` or so to check.

File: src/units/attack_type.cpp
Comment (newfrenchy83) [reply src/units/attack_type.cpp:106]: if self_ exist in this function(lua check) then what not alignment attribute valid, then unit alignment can be returned and second check will use an directly alignment, else alignment is empty and second check should be used in alignment() but i see your point if attribute empty or not valid it's unuse what alignment() make a complete chek and must return unit alignment in case of empty value, i will correct that.

Comment (newfrenchy83) [reply src/units/attack_type.cpp:106]: aalignment() can't have self_ in lua ccheck then i must doing that in init() for what attack alignment function perfect condition and what unit alignment remain value by default.

**Pull request merged**
